### PR TITLE
#6159: open and close mods tab

### LINKED ISSUE #6159: Slice 2: Open/Close Mods Tab
- Update `sidebarSlice.ts` to store open/closed tabs as a `Record<string, boolean>`
    - `state.sidebar.closedTabs`
        - Store using the eventKey to make it easy to determine which tabs to show later
        - `state.sidebar.closedTabs["static-mod-launcher-panel"] = true`
    - Will only store if the mods tab is open/closed to start, but in a later slice, we’ll update it to store the open state of persistent tabs
- Add a `CloseButton` to the staticPanel `TabWithDivider`
    - clicking close closes the tab if there are other tabs open
    - if there are no other tabs, close the sidebar instead
- Add the `+` button to the far right of the open tabs
    - If the modLauncher is closed, clicking opens the modLauncher
    - If the modLauncher is open but not focused, clicking focuses the modLauncher
    - If the modLauncher is open and focused, clicking does nothing
- If there are no other sidebarPanels, the modLauncher should always be visible

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Closes #6159 

## Demo
https://www.loom.com/share/cb271075ec624349992de9c44b30a6af

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer @mnholtz

### CODE REVIEW

File: src/sidebar/Tabs.tsx [hunk 1]
```diff
@@ -121,6 +85,8 @@ const Tabs: React.FC = () => {
   const modActivationPanel = useSelector(selectSidebarModActivationPanel);
   const staticPanels = useSelector(selectSidebarStaticPanels);
   const getExtensionFromEventKey = useSelector(selectExtensionFromEventKey);
+  const closedTabs = useSelector(selectClosedTabs);
+  const hasModLauncher = flagOn("sidebar-home-tab");
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:89]: nit: `hasModLauncherEnabled`

File: src/sidebar/Tabs.module.scss [hunk 2]
```diff
@@ -180,3 +180,20 @@
   text-overflow: ellipsis;
   min-width: 0; // Avoid stretching the flex parent
 }
+
+.addButton {
```
Comment (mnholtz) [inline src/sidebar/Tabs.module.scss:184]: Style nit - the border radius of the button doesn't match the border radius of the other tabs:
Screen Shot 2023-08-04 at 10 34 58 AM

File: src/sidebar/Tabs.tsx [hunk 3]
```diff
@@ -214,15 +218,38 @@ const Tabs: React.FC = () => {
...
+                  onClick={async (event) =>
+                    onCloseStaticPanel(event, staticPanel)
+                  }
+                />
+              </TabWithDivider>
+            );
+          })}
+
+          {hasModLauncher && (
+            <Button
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:243]: I wouldn't have expected just dropping a Button here to play nicely with the bootstrap nav items... but it looks nice 👍

File: src/sidebar/Tabs.tsx [hunk 4]
```diff
@@ -214,15 +218,38 @@ const Tabs: React.FC = () => {
...
 
-          {staticPanels.map((staticPanel) => (
-            <TabWithDivider
-              key={staticPanel.key}
-              active={isPanelActive(staticPanel)}
-              eventKey={eventKeyForEntry(staticPanel)}
+          {staticPanels.map((staticPanel) => {
+            const eventKey = eventKeyForEntry(staticPanel);
+            // eslint-disable-next-line security/detect-object-injection -- checked for eventKey
+            const hidePanel = eventKey in closedTabs && closedTabs[eventKey];
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:224]: What do you think about calling this `isPanelHidden`? It sounds like this is an action/function unless you notice that it's a boolean

File: src/sidebar/Tabs.tsx [hunk 5]
```diff
@@ -135,6 +101,44 @@ const Tabs: React.FC = () => {
...
+
+    // The mod launcher is a special case where we want to hide the sidebar if all other tabs are closed
+    if (
+      staticPanel.key === MOD_LAUNCHER.key &&
+      panels.every((panel) => closedTabs[eventKeyForEntry(panel)])
+    ) {
+      const topLevelFrame = await getTopLevelFrame();
+      await hideSidebar(topLevelFrame);
+    } else {
+      dispatch(sidebarSlice.actions.closeTab(eventKeyForEntry(staticPanel)));
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:138]: What would happen if the user closed the ModLauncher first, then closed all tabs in the sidebar? What would the sidebar look like in that case?

File: src/sidebar/Tabs.tsx [hunk 6]
```diff
@@ -135,6 +101,44 @@ const Tabs: React.FC = () => {
     dispatch(sidebarSlice.actions.removeTemporaryPanel(nonce));
   };
 
+  const onOpenModLauncher = () => {
+    const modLauncherEventKey = eventKeyForEntry(MOD_LAUNCHER);
+    const isModLauncherOpen =
+      // eslint-disable-next-line security/detect-object-injection -- checked for modLauncherEventKey
+      !(modLauncherEventKey in closedTabs) || !closedTabs[modLauncherEventKey];
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:108]: Why is closedTabs being checked twice with `in` and then with the property access?

File: src/sidebar/Tabs.tsx [hunk 7]
```diff
@@ -45,74 +50,33 @@ import {
...
-      </ErrorBoundary>
-    </Tab.Pane>
-  );
-});
-TemporaryPanelTabPane.displayName = "TemporaryPanelTabPane";
-
 const TabWithDivider = ({ children, active, ...props }: NavLinkProps) => (
   <Nav.Item className={cx(styles.tabWrapper, { [styles.active]: active })}>
-    <Nav.Link {...props} className={styles.tabHeader}>
+    <Nav.Link {...props} className={styles.tabHeader} target="_self">
```
Comment (mnholtz) [inline src/sidebar/Tabs.tsx:71]: What is this change about?

File: src/sidebar/sidebar.html [hunk 8]
```diff
@@ -20,7 +20,7 @@
   <head>
     <!--Uncomment this script tag to enable the React Dev Tools app-->
     <!--See https://github.com/pixiebrix/pixiebrix-extension/wiki/Development-commands#react-dev-tools -->
-    <!--<script src="http://localhost:8097"></script>-->
+    <!-- <script src="http://localhost:8097"></script> -->
```
Comment (mnholtz) [inline src/sidebar/sidebar.html:23]: Super tiny nit: I'd keep the comment spacing consistent in the file. Looks like it isn't right now haha

File: src/sidebar/Tabs.tsx (hunk 6 above)
Comment (mnholtz) [reply src/sidebar/Tabs.tsx:108]: The value of closedTabs[key] is a boolean, or undefined if `key` doesn't exist. Would it suffice to just have `!closedTabs[key]`?

File: src/sidebar/Tabs.module.scss (hunk 2 above)
Comment (mnholtz) [reply src/sidebar/Tabs.module.scss:184]: Also, I'd make the bottom-right border radius 0 and expand the button to fill the space to match what the other tabs look like:
Screen Shot 2023-08-04 at 10 34 58 AM

Comment (mnholtz) [reply src/sidebar/Tabs.module.scss:184]: Also, it looks like the hover background color of the button is different from the hover bg color of the tabs? I don't know what the mockup had.
Screen Shot 2023-08-04 at 10 48 47 AM

Comment (grahamlangford) [reply src/sidebar/Tabs.module.scss:184]: Good callout, though I'm not sure it should be the same radius as the tabs. Maybe the radius of the close buttons?

File: src/sidebar/Tabs.test.tsx [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+import { MOD_LAUNCHER } from "@/sidebar/modLauncher/ModLauncher";
+import useFlags from "@/hooks/useFlags";
+import { waitForEffect } from "@/testUtils/testHelpers";
+import userEvent from "@testing-library/user-event";
+
+jest.mock("@/hooks/useFlags", () =>
+  jest.fn().mockReturnValue({ flagOn: jest.fn() })
+);
+
+function setupPanelsAndRender(
```
Comment (mnholtz) [inline src/sidebar/Tabs.test.tsx:34]: ❤️

**Review by mnholtz — COMMENTED** [review]
I have a question to confirm function of the empty state if the Mod Launcher is closed before closing other tabs, and style nits. But logic so far looks good

File: src/sidebar/Tabs.tsx (hunk 5 above)
Comment (grahamlangford) [reply src/sidebar/Tabs.tsx:138]: That's a good question. We can't close the persistent panels yet, that's in a follow-on ticket. We'll probably have to add a check for the equivalent onClose. I'll add a note to check for it in that ticket.

**Review by mnholtz — APPROVED** [review]
I think this is ok to merge with optional nits I mentioned above. We can always iterate on the style later.

File: src/sidebar/Tabs.tsx (hunk 6 above)
Comment (grahamlangford) [reply src/sidebar/Tabs.tsx:108]: You're right. The extra check is unnecessary.

File: src/sidebar/Tabs.tsx (hunk 7 above)
Comment (grahamlangford) [reply src/sidebar/Tabs.tsx:71]: I meant to leave a comment. I will now.

File: src/sidebar/sidebar.html (hunk 8 above)
Comment (grahamlangford) [reply src/sidebar/sidebar.html:23]: lol. Looks like `Cmd+/` added spaces

**Pull request merged**